File: macos/BirkinNativeApp/Sources/BirkinNativeProtocol/NativeProjectionReducer.swift
struct NativeProjectionEvent {
    let protocolVersion: Int
    let sessionID: String
    let cursor: Int
    let eventID: String
    let type: String
    let timestamp: String
    let actorID: String
    let commandID: String
    let payload: NativeJSONObject
}

enum NativeProjectionReducer {
    private static let panelByEvent = [
        "task.updated": "tasks_runs", "run.updated": "tasks_runs",
        "approval.requested": "approvals", "approval.answered": "approvals",
        "question.requested": "approvals", "question.answered": "approvals",
        "evidence.added": "files_evidence", "file.updated": "files_evidence",
        "session.updated": "sessions_history", "activity.recorded": "activity_logs",
        "receipt.recorded": "activity_logs", "integrity.warning": "activity_logs",
        "command.completed": "activity_logs",
        "cron.updated": "cron", "memory.updated": "memory_skills",
        "skill.updated": "memory_skills", "checkpoint.created": "checkpoints_restore",
        "checkpoint.restored": "checkpoints_restore", "computer.updated": "computer_use",
        "progress.updated": "activity_logs", "tool.started": "activity_logs",
        "tool.completed": "activity_logs", "tool.failed": "activity_logs",
        "notification.requested": "activity_logs",
        "settings.updated": "settings_status", "status.updated": "settings_status",
    ]

    static func reduce(
        _ state: inout NativeProjectionState,
        event: NativeProjectionEvent,
        activeCommands: inout Set<String>,
        interrupted: inout Bool
    ) {
        switch event.type {
        case "command.started": activeCommands.insert(event.commandID)
        case "command.completed", "command.failed":
            if activeCommands.remove(event.commandID) == nil {
                activeCommands.remove("__snapshot_active__")
            }
        case "turn.interrupted": interrupted = true
        case "turn.resumed": interrupted = false
        default: break
        }

        if event.type == "message.user", let text = event.payload.string("text") {
            state.conversation.append(message(event: event, kind: "user_message", text: text))
        } else if event.type == "message.assistant.delta",
                  let text = event.payload.string("text") {
            appendAssistantDelta(text, event: event, state: &state)
        } else if event.type == "message.assistant.completed",
                  let text = event.payload.string("text") {
            let value = message(event: event, kind: "assistant_message", text: text)
            if state.conversation.last?.string("kind") == "assistant_stream" {
                state.conversation[state.conversation.count - 1] = value
            } else {
                state.conversation.append(value)
            }
            append(value, toPanel: "sessions_history", state: &state)
        }
        if event.type == "approval.answered" {
            reconcileApproval(event, state: &state)
        } else if let panelKey = panelByEvent[event.type] {
            append(panelItem(event), toPanel: panelKey, state: &state)
        }
        if event.type == "working_memory.updated" {
            applyWorkingMemory(event.payload, state: &state)
        }
        applyTerminal(event, state: &state)

        state.cursor = event.cursor
        state.composer.canSend = activeCommands.isEmpty
        state.composer.canInterrupt = !activeCommands.isEmpty
        state.composer.canResume = interrupted && activeCommands.isEmpty
    }

    private static func applyTerminal(
        _ event: NativeProjectionEvent,
        state: inout NativeProjectionState
    ) {
        guard let terminalID = event.payload.string("terminal_id") else { return }
        if event.type == "terminal.opened" {
            let lease = NativeRedaction.liveValue(event.payload.string("lease"))
            let terminal = NativeTerminalProjection(
                terminalID: terminalID,
                cwd: event.payload.string("cwd") ?? "",
                screen: "",
                outputSequence: 0,
                state: "running",
                exitStatus: nil,
                columns: 80,
                rows: 24,
                lease: lease,
                readOnly: lease == nil
            )
            if let index = state.terminals.firstIndex(where: { $0.terminalID == terminalID }) {
                state.terminals[index] = terminal
            } else {
                state.terminals.append(terminal)
            }
            return
        }
        guard let index = state.terminals.firstIndex(
            where: { $0.terminalID == terminalID }
        ) else { return }
        switch event.type {
        case "terminal.output":
            guard case .int(let sequence) = event.payload["sequence"],
                  sequence == state.terminals[index].outputSequence + 1,
                  let data = event.payload.string("data") else { return }
            state.terminals[index].appendOutput(data)
            state.terminals[index].outputSequence = sequence
        case "terminal.resized":
            if case .int(let columns) = event.payload["columns"] {
                state.terminals[index].columns = columns
            }
            if case .int(let rows) = event.payload["rows"] {
                state.terminals[index].rows = rows
            }
        case "terminal.exited":
            state.terminals[index].state = "exited"
            state.terminals[index].lease = nil
            state.terminals[index].readOnly = true
            if case .int(let status) = event.payload["exit_status"] {
                state.terminals[index].exitStatus = status
            }
        case "terminal.failed":
            state.terminals[index].state = "failed"
            state.terminals[index].lease = nil
            state.terminals[index].readOnly = true
        default: break
        }
    }

    private static func applyWorkingMemory(
        _ payload: NativeJSONObject,
        state: inout NativeProjectionState
    ) {
        guard case .object(let effective) = payload["working_memory"],
              case .int(let revision) = effective["revision"] else { return }
        var fields = state.workingMemory.fields
        for key in fields.keys {
            guard case .array(let values) = effective[key] else { continue }
            let strings = values.compactMap { value -> String? in
                guard case .string(let text) = value else { return nil }
                return text
            }
            fields[key] = strings
        }
        state.workingMemory.revision = revision
        state.workingMemory.fields = fields
    }

    private static func appendAssistantDelta(
        _ text: String,
        event: NativeProjectionEvent,
        state: inout NativeProjectionState
    ) {
        if let last = state.conversation.last,
           last.string("kind") == "assistant_stream",
           let current = last.string("text") {
            state.conversation[state.conversation.count - 1] = [
                "id": last["id"]!,
                "kind": .string("assistant_stream"),
                "text": .string(current + text),
                "actor_id": last["actor_id"]!,
                "cursor": .int(event.cursor),
            ]
        } else {
            state.conversation.append(
                message(event: event, kind: "assistant_stream", text: text)
            )
        }
    }

    private static func message(
        event: NativeProjectionEvent,
        kind: String,
        text: String
    ) -> NativeJSONObject {
        [
            "id": .string(event.eventID), "kind": .string(kind), "text": .string(text),
            "actor_id": .string(event.actorID), "cursor": .int(event.cursor),
        ]
    }

    private static func append(
        _ item: NativeJSONObject,
        toPanel key: String,
        state: inout NativeProjectionState
    ) {
        guard let index = state.panels.firstIndex(where: { $0.key == key }) else { return }
        state.panels[index].items.append(item)
    }

    private static func reconcileApproval(
        _ event: NativeProjectionEvent,
        state: inout NativeProjectionState
    ) {
        guard let approvalID = event.payload.string("approval_id"),
              let panelIndex = state.panels.firstIndex(where: { $0.key == "approvals" })
        else { return }
        let resolved = panelItem(event)
        guard let itemIndex = state.panels[panelIndex].items.firstIndex(
            where: { $0.string("id") == approvalID }
        ) else {
            state.panels[panelIndex].items.append(resolved)
            return
        }
        state.panels[panelIndex].items[itemIndex] = merging(
            state.panels[panelIndex].items[itemIndex],
            with: resolved,
            replacing: [
                "status", "cursor", "kind", "ui_state", "decided",
                "receipt_ref", "effect", "refusal_code",
            ]
        )
    }

    private static func merging(
        _ original: NativeJSONObject,
        with updates: NativeJSONObject,
        replacing keys: Set<String>
    ) -> NativeJSONObject {
        var result = NativeJSONObject()
        for pair in original.pairs {
            try? result.append(
                key: pair.key,
                value: keys.contains(pair.key) ? updates[pair.key] ?? pair.value : pair.value
            )
        }
        for pair in updates.pairs where result[pair.key] == nil && keys.contains(pair.key) {
            try? result.append(key: pair.key, value: pair.value)
        }
        return result
    }

    private static func panelItem(_ event: NativeProjectionEvent) -> NativeJSONObject {
        let payload = event.payload
        let decision = payload.string("decision")
        let id = payload.string("notification_id") ?? payload.string("approval_id")
            ?? payload.string("question_id")
            ?? payload.string("task_id") ?? payload.string("evidence_id")
            ?? payload.string("checkpoint_id") ?? event.eventID
        let summary = payload.string("summary") ?? payload.string("name") ?? event.type
        let status = payload.string("outcome") ?? payload.string("status")
            ?? (event.type == "approval.answered" ? nil : decision)
            ?? event.type.split(separator: ".").last.map(String.init)!
        let kind = [
            "task.updated": "task", "approval.requested": "approval",
            "approval.answered": "approval", "question.requested": "question",
            "question.answered": "question", "evidence.added": "evidence",
            "checkpoint.created": "checkpoint", "checkpoint.restored": "checkpoint",
            "computer.updated": "computer_use", "receipt.recorded": "receipt",
            "command.completed": "receipt", "integrity.warning": "integrity_warning",
            "notification.requested": "notification",
        ][event.type] ?? "activity"
        let defaultState = defaultUIState(event.type, payload: payload)
        var item: NativeJSONObject = [
            "id": .string(id), "summary": .string(summary), "status": .string(status),
            "cursor": .int(event.cursor), "kind": .string(kind),
            "ui_state": .string(payload.string("ui_state") ?? defaultState),
            "updated_at": .string(event.timestamp),
        ]
        for field in [
            "requester", "description", "category", "target", "expected_impact", "rejection_result",
            "related_evidence", "risk", "expires_at", "receipt_ref", "snapshot_ref",
            "effect", "refusal_code", "source_filename", "destination", "authority_digest",
        ] {
            if let value = payload.string(field), !value.isEmpty {
                try? item.append(key: field, value: .string(value))
            }
        }
        for field in ["sealed", "decided", "overwrite_approved"] {
            if case .bool(let value) = payload[field] {
                try? item.append(key: field, value: .bool(value))
            }
        }
        if event.type == "approval.answered", item["decided"] == nil {
            try? item.append(
                key: "decided",
                value: .bool(approvalAnswerIsResolved(payload.string("outcome")))
            )
        }
        if let receipt = payload.string("receipt") {
            try? item.append(key: "receipt_ref", value: .string(receipt))
        }
        if case .int(let sequence) = payload["computer_sequence"] {
            try? item.append(key: "computer_sequence", value: .int(sequence))
        }
        if case .bool(let preserved) = payload["focus_preserved"] {
            try? item.append(key: "focus_preserved", value: .bool(preserved))
        }
        return item
    }

    private static func defaultUIState(
        _ type: String,
        payload: NativeJSONObject
    ) -> String {
        switch type {
        case "approval.requested", "question.requested", "notification.requested":
            return "action_needed"
        case "approval.answered":
            switch payload.string("outcome") {
            case "approved": return "succeeded"
            case "rejected": return "blocked"
            default: return "failed"
            }
        case "question.answered", "evidence.added", "checkpoint.restored": return "succeeded"
        case "task.updated", "tool.started": return "running"
        case "tool.completed": return "succeeded"
        case "tool.failed": return "failed"
        case "computer.updated": return payload.string("ui_state") ?? "pending"
        default: return "pending"
        }
    }

    private static func approvalAnswerIsResolved(_ outcome: String?) -> Bool {
        switch outcome {
        case "approved", "rejected", "answered_elsewhere": true
        default: false
        }
    }
}

private extension NativeJSONObject {
    func string(_ key: String) -> String? {
        guard case .string(let value) = self[key] else { return nil }
        return value
    }
}

File: windows/BirkinNativeApp/src/Birkin.Native.Protocol/Projection/NativeProjectionReducer.cs
using Birkin.Native.Protocol.Framing;

namespace Birkin.Native.Protocol.Projection;

internal static class NativeProjectionReducer
{
    private const int MaxActivityItems = 100;
    private static readonly IReadOnlyDictionary<string, string> PanelByEvent = new Dictionary<string, string>(StringComparer.Ordinal)
    {
        ["task.updated"] = "tasks_runs",
        ["approval.requested"] = "approvals",
        ["receipt.recorded"] = "activity_logs",
        ["integrity.warning"] = "activity_logs",
        ["command.completed"] = "activity_logs",
        ["progress.updated"] = "activity_logs",
        ["tool.started"] = "activity_logs",
        ["tool.completed"] = "activity_logs",
        ["tool.failed"] = "activity_logs",
        ["notification.requested"] = "activity_logs",
    };

    public static NativeProjectionState Reduce(
        NativeProjectionState state, NativeJsonObject body, HashSet<string> activeCommands)
    {
        var type = String(body, "type");
        var commandId = String(body, "command_id");
        var cursor = Integer(body, "cursor");
        var payload = Object(body, "payload");
        if (type == "command.started")
        {
            _ = activeCommands.Add(commandId);
        }
        else if (type is "command.completed" or "command.failed")
        {
            if (!activeCommands.Remove(commandId))
            {
                _ = activeCommands.Remove("__snapshot_active__");
            }
        }

        var conversation = state.Conversation.Values.ToList();
        var panels = state.Panels.Values.Cast<NativeJsonObject>().ToList();
        var terminals = state.Terminals.Values.Cast<NativeJsonObject>().ToList();
        ApplyConversation(conversation, panels, body, type, payload, cursor);
        if (type == "approval.answered")
        {
            ReconcileApproval(panels, body, payload, cursor);
        }
        else if (PanelByEvent.TryGetValue(type, out var panel))
        {
            AppendPanel(panels, panel, PanelItem(body, type, payload, cursor));
        }
        ApplyTerminal(terminals, type, payload);

        var composer = new NativeJsonObject([
            new("can_send", new NativeJsonBoolean(activeCommands.Count == 0)),
            new("can_interrupt", new NativeJsonBoolean(activeCommands.Count != 0)),
            new("can_resume", new NativeJsonBoolean(false)),
        ]);
        return state.WithBody(Replace(state.ToBody(),
            ("cursor", new NativeJsonInteger(cursor)),
            ("panels", new NativeJsonArray(panels)),
            ("conversation", new NativeJsonArray(conversation)),
            ("composer", composer),
            ("terminals", new NativeJsonArray(terminals))));
    }

    private static void ApplyConversation(
        List<NativeJsonValue> conversation,
        List<NativeJsonObject> panels,
        NativeJsonObject body,
        string type,
        NativeJsonObject payload,
        long cursor)
    {
        var text = OptionalString(payload, "text");
        if (text is null)
        {
            return;
        }
        if (type == "message.user")
        {
            conversation.Add(Message(body, "user_message", text, cursor));
        }
        else if (type == "message.assistant.delta")
        {
            if (conversation.LastOrDefault() is NativeJsonObject last
                && OptionalString(last, "kind") == "assistant_stream")
            {
                conversation[^1] = new NativeJsonObject([
                    new("id", last["id"]!),
                    new("kind", new NativeJsonString("assistant_stream")),
                    new("text", new NativeJsonString(String(last, "text") + text)),
                    new("actor_id", last["actor_id"]!),
                    new("cursor", new NativeJsonInteger(cursor)),
                ]);
            }
            else
            {
                conversation.Add(Message(body, "assistant_stream", text, cursor));
            }
        }
        else if (type == "message.assistant.completed")
        {
            var message = Message(body, "assistant_message", text, cursor);
            if (conversation.LastOrDefault() is NativeJsonObject last
                && OptionalString(last, "kind") == "assistant_stream")
            {
                conversation[^1] = message;
            }
            else
            {
                conversation.Add(message);
            }
            AppendPanel(panels, "sessions_history", message);
        }
    }

    private static NativeJsonObject Message(
        NativeJsonObject body,
        string kind,
        string text,
        long cursor) => new([
            new("id", body["event_id"]!),
            new("kind", new NativeJsonString(kind)),
            new("text", new NativeJsonString(text)),
            new("actor_id", body["actor_id"]!),
            new("cursor", new NativeJsonInteger(cursor)),
        ]);

    private static NativeJsonObject PanelItem(
        NativeJsonObject body,
        string type,
        NativeJsonObject payload,
        long cursor)
    {
        var id = OptionalString(payload, "approval_id")
            ?? OptionalString(payload, "task_id")
            ?? OptionalString(payload, "notification_id")
            ?? String(body, "event_id");
        var status = OptionalString(payload, "outcome")
            ?? OptionalString(payload, "status")
            ?? (type == "approval.answered"
                ? null
                : OptionalString(payload, "decision"))
            ?? type[(type.LastIndexOf(".", StringComparison.Ordinal) + 1)..];
        var kind = type switch
        {
            "task.updated" => "task",
            "approval.requested" or "approval.answered" => "approval",
            "receipt.recorded" or "command.completed" => "receipt",
            "integrity.warning" => "integrity_warning",
            "notification.requested" => "notification",
            _ => "activity",
        };
        var projectedUiState = OptionalString(payload, "ui_state");
        var uiState = IsUiState(projectedUiState)
            ? projectedUiState!
            : type switch
            {
                "task.updated" => "running",
                "approval.requested" => "action_needed",
                "approval.answered" => ApprovalAnsweredState(payload),
                "tool.started" => "running",
                "tool.completed" => "succeeded",
                "tool.failed" => "failed",
                "notification.requested" => "action_needed",
                _ => "pending",
            };
        var pairs = new List<KeyValuePair<string, NativeJsonValue>>
        {
            new("id", new NativeJsonString(id)),
            new("summary", new NativeJsonString(OptionalString(payload, "summary") ?? type)),
            new("status", new NativeJsonString(status)),
            new("cursor", new NativeJsonInteger(cursor)),
            new("kind", new NativeJsonString(kind)),
            new("ui_state", new NativeJsonString(uiState)),
        };
        foreach (var field in new[]
        {
            "requester", "description", "category", "target", "expected_impact",
            "rejection_result", "related_evidence", "risk", "issued_at", "expires_at",
            "receipt_ref", "snapshot_ref", "effect", "refusal_code", "session_id",
            "name", "destination", "source_filename", "authority_digest",
            "office_phase", "job_id", "validation_summary",
            "visual_validation_summary",
        })
        {
            if (OptionalString(payload, field) is { Length: > 0 } value)
            {
                pairs.Add(new(field, new NativeJsonString(value)));
            }
        }
        if (type == "receipt.recorded")
        {
            foreach (var field in new[]
            {
                "approval_id", "artifact_id", "draft_id", "diff_id",
                "request_command_id", "approval_command_id",
            })
            {
                if (OptionalString(payload, field) is { Length: > 0 } value)
                {
                    pairs.Add(new(field, new NativeJsonString(value)));
                }
            }
        }
        foreach (var field in new[]
        {
            "sealed", "decided", "overwrite_approved", "backup_exists",
        })
        {
            if (payload[field] is NativeJsonBoolean value)
            {
                pairs.Add(new(field, value));
            }
        }
        if (type == "approval.answered")
        {
            if (!pairs.Any(pair => pair.Key == "decided"))
            {
                pairs.Add(new(
                    "decided",
                    new NativeJsonBoolean(ApprovalAnswerIsResolved(payload))));
            }
            if (!pairs.Any(pair => pair.Key == "receipt_ref")
                && OptionalString(payload, "receipt") is { Length: > 0 } receipt)
            {
                pairs.Add(new("receipt_ref", new NativeJsonString(receipt)));
            }
        }
        return new NativeJsonObject(pairs);
    }

    private static string ApprovalAnsweredState(NativeJsonObject payload) =>
        OptionalString(payload, "outcome") switch
        {
            "approved" => "succeeded",
            "rejected" => "blocked",
            _ => "failed",
        };

    private static bool ApprovalAnswerIsResolved(NativeJsonObject payload) =>
        OptionalString(payload, "outcome") is
            "approved" or "rejected" or "answered_elsewhere";

    private static void ReconcileApproval(
        List<NativeJsonObject> panels,
        NativeJsonObject body,
        NativeJsonObject payload,
        long cursor)
    {
        var approvalId = OptionalString(payload, "approval_id");
        var panelIndex = panels.FindIndex(
            panel => OptionalString(panel, "key") == "approvals");
        if (approvalId is null
            || panelIndex < 0
            || panels[panelIndex]["items"] is not NativeJsonArray items)
        {
            return;
        }
        var resolved = PanelItem(body, "approval.answered", payload, cursor);
        var values = items.Values.ToList();
        var itemIndex = values.FindIndex(
            item => item is NativeJsonObject approval
                && OptionalString(approval, "id") == approvalId);
        if (itemIndex < 0)
        {
            values.Add(resolved);
        }
        else
        {
            values[itemIndex] = Merge(
                (NativeJsonObject)values[itemIndex],
                resolved,
                [
                    "status",
                    "cursor",
                    "kind",
                    "ui_state",
                    "decided",
                    "receipt_ref",
                    "destination",
                    "issued_at",
                    "expires_at",
                    "backup_exists",
                    "validation_summary",
                    "visual_validation_summary",
                    "effect",
                    "refusal_code",
                ]);
        }
        panels[panelIndex] = Replace(
            panels[panelIndex],
            ("items", new NativeJsonArray(values)));
    }

    private static NativeJsonObject Merge(
        NativeJsonObject original,
        NativeJsonObject updates,
        HashSet<string> replacementKeys)
    {
        var result = new List<KeyValuePair<string, NativeJsonValue>>();
        foreach (var pair in original.Pairs)
        {
            var value = replacementKeys.Contains(pair.Key)
                ? updates[pair.Key] ?? pair.Value
                : pair.Value;
            result.Add(new(pair.Key, value));
        }
        foreach (var pair in updates.Pairs)
        {
            if (replacementKeys.Contains(pair.Key)
                && result.All(existing => existing.Key != pair.Key))
            {
                result.Add(pair);
            }
        }
        return new NativeJsonObject(result);
    }

    private static void AppendPanel(
        List<NativeJsonObject> panels,
        string key,
        NativeJsonObject item)
    {
        var index = panels.FindIndex(panel => OptionalString(panel, "key") == key);
        if (index < 0 || panels[index]["items"] is not NativeJsonArray items)
        {
            return;
        }
        var values = items.Values.Append(item);
        if (string.Equals(key, "activity_logs", StringComparison.Ordinal))
        {
            values = values.TakeLast(MaxActivityItems);
        }
        panels[index] = Replace(
            panels[index],
            ("items", new NativeJsonArray(values)));
    }

    private static bool IsUiState(string? value) => value is
        "action_needed"
        or "blocked"
        or "failed"
        or "paused"
        or "pending"
        or "running"
        or "succeeded";

    private static void ApplyTerminal(
        List<NativeJsonObject> terminals, string type, NativeJsonObject payload)
    {
        var terminalId = OptionalString(payload, "terminal_id");
        if (terminalId is null)
        {
            return;
        }
        var index = terminals.FindIndex(item => OptionalString(item, "terminal_id") == terminalId);
        if (type == "terminal.opened")
        {
            var terminal = new NativeJsonObject([
                new("terminal_id", new NativeJsonString(terminalId)),
                new("cwd", new NativeJsonString(OptionalString(payload, "cwd") ?? string.Empty)),
                new("screen", new NativeJsonString(string.Empty)),
                new("output_sequence", new NativeJsonInteger(0)),
                new("state", new NativeJsonString("running")),
                new("exit_status", NativeJsonNull.Value),
                new("columns", new NativeJsonInteger(80)),
                new("rows", new NativeJsonInteger(24)),
                new("lease", NativeJsonNull.Value),
                new("read_only", new NativeJsonBoolean(true)),
            ]);
            if (index < 0)
            {
                terminals.Add(terminal);
            }
            else
            {
                terminals[index] = terminal;
            }
            return;
        }
        if (index < 0)
        {
            return;
        }
        var current = terminals[index];
        terminals[index] = type switch
        {
            "terminal.output" => Replace(current,
                ("screen", new NativeJsonString(String(current, "screen") + (OptionalString(payload, "data") ?? string.Empty))),
                ("output_sequence", payload["sequence"] ?? current["output_sequence"]!)),
            "terminal.resized" => Replace(current,
                ("columns", payload["columns"] ?? current["columns"]!),
                ("rows", payload["rows"] ?? current["rows"]!)),
            "terminal.exited" => Replace(current,
                ("state", new NativeJsonString("exited")),
                ("exit_status", payload["exit_status"] ?? NativeJsonNull.Value),
                ("lease", NativeJsonNull.Value),
                ("read_only", new NativeJsonBoolean(true))),
            _ => current,
        };
    }

    private static NativeJsonObject Replace(
        NativeJsonObject source,
        params (string Key, NativeJsonValue Value)[] replacements)
    {
        var values = replacements.ToDictionary(pair => pair.Key, pair => pair.Value, StringComparer.Ordinal);
        return new NativeJsonObject(source.Pairs.Select(pair =>
            new KeyValuePair<string, NativeJsonValue>(pair.Key, values.GetValueOrDefault(pair.Key, pair.Value))));
    }

    private static NativeJsonObject Object(NativeJsonObject body, string key) =>
        body[key] as NativeJsonObject ?? throw BodyError();

    private static long Integer(NativeJsonObject body, string key) =>
        body[key] is NativeJsonInteger integer ? integer.Value : throw BodyError();

    private static string String(NativeJsonObject body, string key) =>
        OptionalString(body, key) ?? throw BodyError();

    private static string? OptionalString(NativeJsonObject body, string key) =>
        body[key] is NativeJsonString text ? text.Value : null;

    private static NativeProtocolError BodyError() => new("E_BODY", "projection event body is invalid");
}
